runtime/cudaq/platform/mqpu/remote/RemoteSimulatorQPU.cpp

```diff
@@ -0,0 +1,97 @@
+ * Copyright (c) 2022 - 2023 NVIDIA Corporation & Affiliates.                  *
+ * All rights reserved.                                                        *
+ *                                                                             *
+ * This source code and the accompanying materials are made available under    *
+ * the terms of the Apache License 2.0 which accompanies this distribution.    *
+ ******************************************************************************/
+
+#include "common/ExecutionContext.h"
+#include "common/Logger.h"
+#include "common/RemoteKernelExecutor.h"
+#include "common/RuntimeMLIR.h"
+#include "cudaq.h"
+#include "cudaq/platform/qpu.h"
+#include "cudaq/platform/quantum_platform.h"
+
+namespace {
+using namespace mlir;
+
+// Remote QPU: delegating the execution to a remotely-hosted server, which can
+// reinstate the execution context and JIT-invoke the kernel.
+class RemoteSimulatorQPU : public cudaq::QPU {
+private:
+  std::string m_simName;
+  std::unordered_map<std::thread::id, cudaq::ExecutionContext *> m_contexts;
+  std::mutex m_contextMutex;
+  std::unique_ptr<MLIRContext> m_mlirContext;
+  std::unique_ptr<cudaq::RemoteRuntimeClient> m_client;
+
+public:
+  RemoteSimulatorQPU()
+      : QPU(), m_mlirContext(cudaq::initializeMLIR()),
+        m_client(cudaq::registry::get<cudaq::RemoteRuntimeClient>("rest")) {}
+
+  std::thread::id getExecutionThreadId() const {
+    return execution_queue->getExecutionThreadId();
+  }
+
+  virtual void setTargetBackend(const std::string &backend) override {
+    auto parts = cudaq::split(backend, ';');
+    if (parts.size() % 2 != 0)
+      throw std::invalid_argument("Unexpected backend configuration string. "
+                                  "Expecting a ';'-separated key-value pairs.");
+    for (std::size_t i = 0; i < parts.size(); i += 2) {
+      if (parts[i] == "url")
+        m_client->setConfig({{"url", parts[i + 1]}});
+      if (parts[i] == "simulator")
+        m_simName = parts[i + 1];
+    }
+  }
+
+  void enqueue(cudaq::QuantumTask &task) override {
+    cudaq::info("RemoteSimulatorQPU: Enqueue Task on QPU {}", qpu_id);
+    execution_queue->enqueue(task);
+  }
+
+  void launchKernel(const std::string &name, void (*kernelFunc)(void *),
+                    void *args, std::uint64_t voidStarSize,
+                    std::uint64_t resultOffset) override {
+    cudaq::info("RemoteSimulatorQPU: Launch kernel named '{}' remote QPU {} "
+                "(simulator = {})",
+                name, qpu_id, m_simName);
+
+    cudaq::ExecutionContext *executionContextPtr =
+        [&]() -> cudaq::ExecutionContext * {
+      std::scoped_lock<std::mutex> lock(m_contextMutex);
+      const auto iter = m_contexts.find(std::this_thread::get_id());
+      if (iter == m_contexts.end())
+        return nullptr;
+      return iter->second;
+    }();
+    static cudaq::ExecutionContext defaultContext("sample", 1);
```

Comment:
Is there any chance that multiple threads would try to use `defaultContext`? If not, perhaps add a comment explaining why that is. (It's a little surprising that the above lambda needs a lock to access `m_contexts` but this usage of `defaultContext` wouldn't need one.

Also:
```suggestion
    static cudaq::ExecutionContext defaultContext("sample", /*shots=*/1);
```

Reply:
It's very unlikely since this is really an edge case. Anyway, I've changed this to be a `thread_local` along with code comment for why we might need this dummy variable (https://github.com/NVIDIA/cuda-quantum/pull/1012/commits/4b8237322f21415f53f19b9367e08d997fcc4783).